drivers/modem/quectel-bg9x.c

```diff
@@ -0,0 +1,1147 @@
+ * up and kill the driver.
+ */
+static int modem_at(struct modem_context *mctx, struct modem_data *mdata)
+{
+	int counter = 0, ret = -1;
+
+	while (counter < 50 && ret < 0) {
+
+		/* Send "AT" command to the modem. */
+		ret = modem_cmd_send(&mctx->iface, &mctx->cmd_handler,
+				     NULL, 0, "AT", &mdata->sem_response,
+				     MDM_CMD_TIMEOUT);
+
+		/* Check the response from the Modem. */
+		if (ret < 0 && ret != -ETIMEDOUT) {
+			return ret;
+		}
+
+		counter++;
+		k_sleep(K_SECONDS(2));
+	}
+
+	return ret;
+}
+
+/* Func: on_cmd_sockread_common
+ * Desc: Function to successfully read data from the modem on a given socket.
+ */
+static int on_cmd_sockread_common(int socket_fd,
+				  struct modem_cmd_handler_data *data,
+				  uint16_t len)
+{
+	struct modem_socket	 *sock = NULL;
+	struct socket_read_data	 *sock_data;
+	int	ret, i;
+	int socket_data_length = find_len(data->rx_buf->data);
+	int bytes_to_skip;
+
+	if (!len) {
+		LOG_ERR("Invalid length, Aborting!");
+		return -EAGAIN;
+	}
+
+	/* Make sure we still have buf data */
+	if (!data->rx_buf) {
+		LOG_ERR("Incorrect format! Ignoring data!");
+		return -EINVAL;
+	}
+
+	/* No (or not enough) data available on the socket. */
+	bytes_to_skip = digits(socket_data_length) + 2 + 4;
+	if (socket_data_length <= 0) {
+		LOG_ERR("Length problem (%d).  Aborting!", socket_data_length);
+		return -EAGAIN;
+	}
+
+	/* check to make sure we have all of the data. */
+	if (net_buf_frags_len(data->rx_buf) < (socket_data_length + bytes_to_skip)) {
+		LOG_DBG("Not enough data -- wait!");
+		return -EAGAIN;
+	}
+
+	/* Skip "len" and CRLF */
+	bytes_to_skip = digits(socket_data_length) + 2;
+	for (i = 0; i < bytes_to_skip; i++) {
+		net_buf_pull_u8(data->rx_buf);
+	}
+
+	if (!data->rx_buf->len) {
+		data->rx_buf = net_buf_frag_del(NULL, data->rx_buf);
+	}
+
+	sock = modem_socket_from_fd(&mdata.socket_config, socket_fd);
+	if (!sock) {
+		LOG_ERR("Socket not found! (%d)", socket_fd);
+		ret = -EINVAL;
+		goto exit;
+	}
+
+	sock_data = (struct socket_read_data *)sock->data;
+	if (!sock_data) {
+		LOG_ERR("Socket data not found! Skip handling (%d)", socket_fd);
+		ret = -EINVAL;
+		goto exit;
+	}
+
+	ret = net_buf_linearize(sock_data->recv_buf, sock_data->recv_buf_len,
+				data->rx_buf, 0, (uint16_t)socket_data_length);
+	data->rx_buf = net_buf_skip(data->rx_buf, ret);
+	sock_data->recv_read_len = ret;
+	if (ret != socket_data_length) {
+		LOG_ERR("Total copied data is different then received data!"
+			" copied:%d vs. received:%d", ret, socket_data_length);
+		ret = -EINVAL;
+	}
+
+exit:
+	/* remove packet from list (ignore errors) */
+	(void)modem_socket_packet_size_update(&mdata.socket_config, sock,
+					      -socket_data_length);
+
+	/* don't give back semaphore -- OK to follow */
+	return ret;
+}
+
+/* Func: socket_close
+ * Desc: Function to close the given socket descriptor.
+ */
+static void socket_close(struct modem_socket *sock)
+{
+	char buf[32];
```

Comment:
Instead of hard coded value, we could use a template for expected input here. So something like
`char buf[sizeof("AT+QICLOSE=xx")];`
assuming here that the sock_fd is <100.